library/src/blas3/rocblas_herk.hpp

```diff
@@ -0,0 +1,260 @@
+    int to   = upper ? ty : tx;
+
+    if(tx < n && ty < n)
+    {
+        if(from < to)
+        {
+            C[ty * ldc + tx] *= beta;
+        }
+        else if(from == to)
+        {
+            // multiply only real component and zero imaginary on diagonal
+            U& e = C[ty * ldc + tx];
+            e.x  = beta * e.x;
+            e.y  = 0.0;
+        }
+    }
+}
+
+/**
+  *  Loads pointers and launches the actual calculation kernel.
+  */
+template <typename U, typename V>
+__global__ void herk_scale_kernel(bool           upper,
+                                  rocblas_int    n,
+                                  U              alpha_host_device,
+                                  U              beta_host_device,
+                                  V              CP_array,
+                                  ptrdiff_t      shift_c,
+                                  rocblas_int    ldc,
+                                  rocblas_stride stride_c)
+{
+
+    auto C     = load_ptr_batch(CP_array, hipBlockIdx_z, shift_c, stride_c);
+    auto alpha = load_scalar(alpha_host_device);
+    auto beta  = load_scalar(beta_host_device);
+
+    if(beta == 1 && alpha == 0) // if alpha not zero we need imaginary clear on diagonal
+        return;
+
+    herk_scale_device(upper, n, beta, C, ldc);
+}
+
+template <typename TScal, typename TConstPtr, typename TPtr>
+rocblas_status rocblas_herk_arg_check(rocblas_handle    handle,
+                                      rocblas_fill      uplo,
+                                      rocblas_operation transA,
+                                      rocblas_int       n,
+                                      rocblas_int       k,
+                                      TScal             alpha,
+                                      TConstPtr         AP,
+                                      rocblas_int       offsetA,
+                                      rocblas_int       lda,
+                                      rocblas_stride    strideA,
+                                      TScal             beta,
+                                      TPtr              CP,
+                                      rocblas_int       offsetC,
+                                      rocblas_int       ldc,
+                                      rocblas_stride    strideC,
+                                      rocblas_int       batch_count)
+{
+    if(uplo != rocblas_fill_lower && uplo != rocblas_fill_upper)
+        return rocblas_status_invalid_value;
+    if(transA != rocblas_operation_none && transA != rocblas_operation_conjugate_transpose)
+        return rocblas_status_invalid_value;
+
+    if(n < 0 || k < 0 || batch_count < 0 || ldc < n || (transA == rocblas_operation_none && lda < n)
+       || (transA != rocblas_operation_none && lda < k))
+        return rocblas_status_invalid_size;
+    if(!n || !batch_count)
+        return rocblas_status_success;
+    if(!AP || !CP || !alpha || !beta)
```

Comment:
I recall for gemm coming across a jira ticket which brought up the issue of "if k == 0, do we care if A is a nullptr?" I'll forward you the email chain we had.
I think that for gemm, if k == 0, A is allowed to be a nullptr. I'm not sure if we should adopt that standard library-wide or if this can be ignored for now. We can probably talk about it at the meeting on Monday.

Reply:
That makes sense I can add if (k > 0 && (!AP || !alpha) || ...

Reply:
This will be examined when I do the sweep of the argument error-checking across all routines.